Mathlib/Algebra/MonoidAlgebra/NoZeroDivisors.lean

```diff
@@ -161,4 +190,95 @@ theorem NoZeroDivisors.of_right_ordered [NoZeroDivisors R] [AddLeftCancelSemigro
+  refine Finsupp.not_mem_support_iff.mp (fun h ↦ ?_)
+  refine not_not.mpr ((support_mul (Finsupp.single a r) (f.erase t)) h) ?_
+  simp only [Finsupp.mem_support_single, Finset.mem_biUnion, Ne.def, Finset.mem_singleton,
+    exists_prop, not_exists, not_and, and_imp, forall_eq, Finsupp.support_erase]
+  intros _ x xs
+  refine (add_lt_add_left (WithBot.coe_lt_coe.mp ?_) _).ne'
+  refine WithBot.coe_lt_coe.mpr ((ft _ (Finset.mem_of_mem_erase xs)).lt_of_ne ?_)
+  exact Finset.ne_of_mem_erase xs
+
+/--  The "bottom" element of `Finsupp.single a r * f` is the product of `r` and
+the "bottom" element of `f`.  Here, "bottom" is simply a lower bound for the elements
+of the support of `f` (e.g. the product is `0` if "bottom" is not a minimum).
+The corresponding statement for a general product `f * g` is `AddMonoidAlgebra.mul_apply_of_le'`.
+It is proved with a further `CovariantClass` assumption. -/
+theorem single_mul_apply_of_le' (r : R) (fb : ∀ a ∈ f.support, b ≤ a) :
+  ((AddMonoidAlgebra.single a r) * f) (a + b) = r * f b :=
+@single_mul_apply_of_le _ Aᵒᵈ _ _ _ _ _ _ _ _ fb
+
+variable [CovariantClass A A (Function.swap (· + ·)) (· < ·)]
+
+/--  The "top" element of `f * g` is the product of the "top" elements of `f` and of `g`.
+Here, "top" is simply an upper bound for the elements of the support of the corresponding
+polynomial (e.g. the product is `0` if "top" is not a maximum). -/
+theorem mul_apply_of_le (fa : ∀ i ∈ f.support, i ≤ a) (gt : ∀ i ∈ g.support, i ≤ t) :
+  (f * g) (a + t) = f a * g t := by
+  classical
+  nth_rw 1 [← f.erase_add_single a]
+  rw [add_mul, Finsupp.add_apply, single_mul_apply_of_le _ gt]
+  convert zero_add _
+  refine Finsupp.not_mem_support_iff.mp (fun h ↦ ?_)
+  refine not_not.mpr ((support_mul _ g) h) ?_
+  simp only [Finsupp.support_erase, Finset.mem_biUnion, Finset.mem_erase, Ne.def,
+    Finset.mem_singleton, exists_prop, not_exists, not_and, and_imp]
+  haveI : CovariantClass A A (· + ·) (· ≤ ·) := Add.to_covariantClass_left A
+  exact fun x xa xf y yg ↦ (add_lt_add_of_lt_of_le ((fa _ xf).lt_of_ne xa) (gt _ yg)).ne'
+
+/--  The "bottom" element of `f * g` is the product of the "bottom" elements of `f` and of `g`.
+Here, "bottom" is simply a lower bound for the elements of the support of the corresponding
+polynomial (e.g. the product is `0` if "bottom" is not a minimum). -/
+theorem mul_apply_of_le' (fa : ∀ i ∈ f.support, a ≤ i) (gb : ∀ i ∈ g.support, b ≤ i) :
+  (f * g) (a + b) = f a * g b :=
+@mul_apply_of_le _ Aᵒᵈ _ _ _ _ _ _ _ _ _ fa gb
+
+theorem mul_apply_eq_zero_of_lt (fa : ∀ i ∈ f.support, a ≤ i) (gb : ∀ i ∈ g.support, b < i) :
+  (f * g) (a + b) = 0 := by
+  rw [mul_apply_of_le' fa (fun x hx ↦ ((gb _ hx).le))]
+  convert mul_zero _
+  exact Finsupp.not_mem_support_iff.mp (fun bg ↦ (lt_irrefl b (gb b bg)).elim)
+
+end PartialOrder
+
+section LinearOrder
+variable [NoZeroDivisors R] [Add A] [LinearOrder A] [CovariantClass A A (· + ·) (· < ·)]
+  [CovariantClass A A (Function.swap (· + ·)) (· < ·)]
```

Comment:
Maybe we could make `UniqueSums` a typeclass and then `NoZeroDivisors.biOrdered` would be `inferInstance`.

Also, it should be trivial to add the `MonoidAlgebra` version, but I think `to_additive` doesn't work.

Reply:
Yes, you are right: the plan was indeed to use `UniqueSums`!  I simply got stuck a few PRs before the port started...  I'll fix it, but now is time to sleep for me!

Reply:
Thanks and no worries! It's my bad that I already realized the PR is outdated  but still mentioned it. No rush to update PR because I find that stuff around `Covariant`/`ContravariantClass`es should benefit from some refactoring utilizing Lean 4's loopy instances: there are lots of results of the shape A ↔ B, or A → B and B ∧ C → A. It would be nice to unify the assumptions of [add_eq_add_iff_eq_and_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/Monoid/Lemmas.html#add_eq_add_iff_eq_and_eq) with your [eq_and_eq_of_le_of_le_of_add_le](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Order/Monoid/Basic.html#eq_and_eq_of_le_of_le_of_add_le), for example. I'll keep you updated!